# **Project: Stage I (Problem and Tasks)**
# COVID-19 Data Analysis
## Description
COVID-19 is a continuing worldwide pandemic which has affected a lot of people, including you. Our goal with the major project in this class is to develop an analytical framework to study the data coming from the United States, to understand patterns of COVID-19 effects and spread.

In order to achieve that, the project is separated into four stages:

- Stage I: Data and Project Understanding
- Stage II: Data Modeling and Hypothesis Testing
- Stage III: Basic Machine Learning
- Stage IV: Dashboard

## Project Stage I: Data and Project Understanding
## COVID-19 Dataset
We'll utilize data from usfacts.org. The dataset contains daily county-level trackers of COVID-19 cases. This makes it easy to follow COVID-19 cases on a granular level, as does the ability to break down infections per 100,000 people (with population data). The underlying data is available for download below the US county map, and has helped government agencies like the Centers for Disease Control and Prevention in its nationwide efforts.

- [USA Facts: US COVID-19 cases and deaths by state](https://usafacts.org/visualizations/coronavirus-covid-19-spread-map/)
- [USA Facts: Number of Cases (.csv file)](https://usafactsstatic.blob.core.windows.net/public/data/covid-19/covid_confirmed_usafacts.csv)
- [USA Facts: Number of Deaths (.csv file)](https://usafactsstatic.blob.core.windows.net/public/data/covid-19/covid_deaths_usafacts.csv)
- [USA Facts: Population by County (.csv file)](https://usafactsstatic.blob.core.windows.net/public/data/covid-19/covid_county_population_usafacts.csv) 

## Task 1
The entire team looks at the COVID-19 dataset and understands the type of variables present in each of the data.

### Team
- Section in the report describing the COVID-19 dataset and datatype: variable dictionary.
- Preliminary intuitions from the data.
- Each member of the team takes on an enrichment dataset. They read the data descriptions and understand the variables present in the data.

### Individual
- Section in the report describing the enrichment data and datatype: variable dictionary.
- How can you merge the data with the primary COVID-19 dataset? Identify the individual variable which maps between the datasets.
- Describe how your enrichment data can help in the analysis of COVID-19 spread. Pose initial hypothesis questions.

In [147]:
import os # operating system
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import string
import random
import datetime
import matplotlib.pyplot as plt
%matplotlib inline

##### We have imported all required libraries for Team task, Covid-19 data analysis

In [148]:
USConfirmCasesDF = pd.read_csv("..\\..\\Data\\covid_confirmed_usafacts.csv")
USDeathsDF = pd.read_csv("..\\..\\Data\\covid_deaths_usafacts.csv")
USCountyPopulationDF = pd.read_csv("..\\..\\Data\\covid_county_population_usafacts.csv")

##### We have downloaded latest data from provided website *USA Facts: US COVID-19 cases and deaths by state* for below datasets. 
###### US County level data available until Feb 10th 2022 for all states
- Covid19 Confirmed Cases
- Covid19 Death Cases 
- County Population 

In [149]:
USConfirmCasesDF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3193 entries, 0 to 3192
Columns: 755 entries, countyFIPS to 2022-02-10
dtypes: int64(753), object(2)
memory usage: 18.4+ MB


##### We see that US Covid19 Confirmed Cases Dataset contains 3193 observations and 755 columns(753 Integers & 2 Object datatypes)

In [150]:
USDeathsDF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3193 entries, 0 to 3192
Columns: 755 entries, countyFIPS to 2022-02-10
dtypes: int64(753), object(2)
memory usage: 18.4+ MB


##### We see that US Covid19 Death Cases Dataset also contains 3193 observations and 755 columns(753 Integers & 2 Object datatypes)

In [151]:
USCountyPopulationDF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3195 entries, 0 to 3194
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   countyFIPS   3195 non-null   int64 
 1   County Name  3195 non-null   object
 2   State        3195 non-null   object
 3   population   3195 non-null   int64 
dtypes: int64(2), object(2)
memory usage: 100.0+ KB


##### We observe that US County Population Dataset contains 3195 observations and 4 columns(2 Integers & 2 Object datatypes). All of these data columns does not contain any NaN, NA values.

In [152]:
USConfirmCasesDF.isnull().sum()

countyFIPS     0
County Name    0
State          0
StateFIPS      0
2020-01-22     0
              ..
2022-02-06     0
2022-02-07     0
2022-02-08     0
2022-02-09     0
2022-02-10     0
Length: 755, dtype: int64

##### We also verified US Coovid19 Confirm Case data has any null columns and didnt find any, above results shows all Null counts are Zero(0)

In [153]:
USDeathsDF.isnull().sum()

countyFIPS     0
County Name    0
State          0
StateFIPS      0
2020-01-22     0
              ..
2022-02-06     0
2022-02-07     0
2022-02-08     0
2022-02-09     0
2022-02-10     0
Length: 755, dtype: int64

##### We also verified US Coovid19 Death data has any null columns and didnt find any, above results shows all Null counts are Zero(0)

In [154]:
USConfirmCasesDF.rename(columns={'countyFIPS': 'CountyFIPS', 'County Name': 'CountyName'}, inplace=True)
USDeathsDF.rename(columns={'countyFIPS': 'CountyFIPS', 'County Name': 'CountyName'}, inplace=True)
USCountyPopulationDF.rename(columns={'countyFIPS': 'CountyFIPS', 'County Name': 'CountyName', 'population': 'Population'}, inplace=True)

##### As per unique naming standards and easy understand code , we have renamed all columns with Init CAPS. Ex: County Name to CountyName, countyFIPS to CountyFIPS

In [155]:
VarIndex = ['State', 'StateFIPS', 'CountyFIPS', 'CountyName']
VarValue = USConfirmCasesDF.columns.difference(VarIndex)
USConfirmCasesLongDF = USConfirmCasesDF.melt(id_vars=VarIndex, value_vars = VarValue, var_name='ReportDate', value_name='ConfirmCasesCount')
USConfirmCasesLongDF = USConfirmCasesLongDF[['ReportDate', 'StateFIPS', 'State', 'CountyFIPS', 'CountyName', 'ConfirmCasesCount']]
USConfirmCasesLongDF.head()

ReportDate  StateFIPS State  CountyFIPS             CountyName  \
0  2020-01-22          1    AL           0  Statewide Unallocated   
1  2020-01-22          1    AL        1001        Autauga County    
2  2020-01-22          1    AL        1003        Baldwin County    
3  2020-01-22          1    AL        1005        Barbour County    
4  2020-01-22          1    AL        1007           Bibb County    

   ConfirmCasesCount  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0

##### We have observe that US Covid19 Confirmed Cases is wide dataset, which needs to make long dataset to combine & use with other datasets.
- Taken all common columns for Indexing into variable VarIndex. i.e. State, StateFIPS, CountyFIPS, CountyName
- And then taken rest of all columns(755-4=751) into another variable VarValue
- We have used Pandas Melt function to convert Pivot to unpivot table by passing VarIndex & Varvalue
- Named New columns as "ReportDate" and "ConfirmCaseCount"
- And also rearranged columns as per standards and easy understand

In [156]:
VarIndex = ['State', 'StateFIPS', 'CountyFIPS', 'CountyName']
VarValue = USDeathsDF.columns.difference(VarIndex)
USDeathsLongDF = USDeathsDF.melt(id_vars=VarIndex, value_vars = VarValue, var_name='ReportDate', value_name='DeathCount')
USDeathsLongDF = USDeathsLongDF[['ReportDate', 'StateFIPS', 'State', 'CountyFIPS', 'CountyName', 'DeathCount']]
USDeathsLongDF.head()

ReportDate  StateFIPS State  CountyFIPS             CountyName  DeathCount
0  2020-01-22          1    AL           0  Statewide Unallocated           0
1  2020-01-22          1    AL        1001        Autauga County            0
2  2020-01-22          1    AL        1003        Baldwin County            0
3  2020-01-22          1    AL        1005        Barbour County            0
4  2020-01-22          1    AL        1007           Bibb County            0

##### We have observe that US Covid19 Death is wide dataset, which needs to make long dataset to combine & use with other datasets.
- Taken all common columns for Indexing into variable VarIndex. i.e. State, StateFIPS, CountyFIPS, CountyName
- And then taken rest of all columns(755-4=751) into another variable VarValue
- We have used Pandas Melt function to convert Pivot to unpivot table by passing VarIndex & Varvalue
- Named New columns as "ReportDate" and "DeathCount"
- And also rearranged columns as per standards and easy understand

In [157]:
USConfirmCasesLongDF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2397943 entries, 0 to 2397942
Data columns (total 6 columns):
 #   Column             Dtype 
---  ------             ----- 
 0   ReportDate         object
 1   StateFIPS          int64 
 2   State              object
 3   CountyFIPS         int64 
 4   CountyName         object
 5   ConfirmCasesCount  int64 
dtypes: int64(3), object(3)
memory usage: 109.8+ MB


##### We have verified after converision of long format datatypes, observations & columns
- Observations increased from 3193 to 2,397,943 rows for each report date, it shows Confirm Case Count by State, StateFIPS, CountyFIPS, CountyName
- Total 6 columns with ReportDate, State, CountyName are Object, StateFIPS, CountyFIPS, ConfirmCaseCount are Integers

In [158]:
USDeathsLongDF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2397943 entries, 0 to 2397942
Data columns (total 6 columns):
 #   Column      Dtype 
---  ------      ----- 
 0   ReportDate  object
 1   StateFIPS   int64 
 2   State       object
 3   CountyFIPS  int64 
 4   CountyName  object
 5   DeathCount  int64 
dtypes: int64(3), object(3)
memory usage: 109.8+ MB


##### We have verified after converision of long format datatypes, observations & columns
- Observations increased from 3193 to 2,397,943 rows for each report date, it shows Confirm Case Count by State, StateFIPS, CountyFIPS, CountyName
- Total 6 columns with ReportDate, State, CountyName are Object, StateFIPS, CountyFIPS, DeathCount are Integers

In [159]:
JoinKeys = ['ReportDate', 'StateFIPS', 'State', 'CountyFIPS', 'CountyName']
USConfirmCasesDeathsDF = pd.merge(USConfirmCasesLongDF, USDeathsLongDF,on=JoinKeys, how='outer')
USConfirmCasesDeathsDF.head()

ReportDate  StateFIPS State  CountyFIPS             CountyName  \
0  2020-01-22          1    AL           0  Statewide Unallocated   
1  2020-01-22          1    AL        1001        Autauga County    
2  2020-01-22          1    AL        1003        Baldwin County    
3  2020-01-22          1    AL        1005        Barbour County    
4  2020-01-22          1    AL        1007           Bibb County    

   ConfirmCasesCount  DeathCount  
0                  0           0  
1                  0           0  
2                  0           0  
3                  0           0  
4                  0           0

##### We have merged US Covid19 Confirmed Cases and Death datasets by using below steps.
- Taken all common columns between data sets. i.e. State, StateFIPS, CountyFIPS, CountyName
- Even though we can just join based on StateFIPS, CountyFIPS, we have taken Names as well since they all come from same soure i.e. usafacts.org
- We have tested with and without names, both gives same results.
- Also to avoid missing data due to joins, we have used Outer join.
- So merged based on above columns and verified data, they are all looks good

In [160]:
USConfirmCasesDeathsDF.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2397943 entries, 0 to 2397942
Data columns (total 7 columns):
 #   Column             Dtype 
---  ------             ----- 
 0   ReportDate         object
 1   StateFIPS          int64 
 2   State              object
 3   CountyFIPS         int64 
 4   CountyName         object
 5   ConfirmCasesCount  int64 
 6   DeathCount         int64 
dtypes: int64(4), object(3)
memory usage: 146.4+ MB


##### We observe that combined data for US Confirm and Deaths data "USConfirmCasesDeathsDF" shows same amount of rows 2,397,943 with 7 columns
-  Total 7 columns with ReportDate, State, CountyName are Object, StateFIPS, CountyFIPS, ConfirmCaseCount, DeathCount are Integers

In [161]:
USConfirmCasesDeathsDF.isnull().sum()

ReportDate           0
StateFIPS            0
State                0
CountyFIPS           0
CountyName           0
ConfirmCasesCount    0
DeathCount           0
dtype: int64

##### We verified data and no nulls exists in all columns.

In [162]:
USCountyPopulationDF.head()

CountyFIPS             CountyName State  Population
0           0  Statewide Unallocated    AL           0
1        1001         Autauga County    AL       55869
2        1003         Baldwin County    AL      223234
3        1005         Barbour County    AL       24686
4        1007            Bibb County    AL       22394

##### Before join population data, we have verified US Population data for all counties.

In [163]:
JoinKeys = ['State', 'CountyFIPS','CountyName']
USConfirmCasesDeathsDF['CountyName'] = USConfirmCasesDeathsDF["CountyName"].str.strip()
USCovid19DF = pd.merge(USConfirmCasesDeathsDF, USCountyPopulationDF, on=JoinKeys, how='left')
USCovid19DF.head()

ReportDate  StateFIPS State  CountyFIPS             CountyName  \
0  2020-01-22          1    AL           0  Statewide Unallocated   
1  2020-01-22          1    AL        1001         Autauga County   
2  2020-01-22          1    AL        1003         Baldwin County   
3  2020-01-22          1    AL        1005         Barbour County   
4  2020-01-22          1    AL        1007            Bibb County   

   ConfirmCasesCount  DeathCount  Population  
0                  0           0         0.0  
1                  0           0     55869.0  
2                  0           0    223234.0  
3                  0           0     24686.0  
4                  0           0     22394.0

##### We have merged US Covid19 Confirmed Death combined and Population datasets by using below steps.
- Taken all common columns between data sets. i.e. State, CountyFIPS, CountyName
- Even though we can just join based on CountyFIPS, we have taken Names as well since they all come from same soure i.e. usafacts.org
- We have tested with and without names, both gives same results.
- Also to avoid missing data due to joins, we have used left join on County Population to ConfirmCase&Deaths.
- So merged based on above columns and verified data, they are all looks good except Population column as float64

In [164]:
#Mapping the state with its Names. This will be helpful later when we perform merging with the super COVID-19 dataset.
USStateMapping = {
    'AK': 'Alaska',
    'AL': 'Alabama',
    'AR': 'Arkansas',
    'AZ': 'Arizona',
    'CA': 'California',
    'CO': 'Colorado',
    'CT': 'Connecticut',
    'DC': 'District of Columbia',
    'DE': 'Delaware',
    'FL': 'Florida',
    'GA': 'Georgia',
    'HI': 'Hawaii',
    'IA': 'Iowa',
    'ID': 'Idaho',
    'IL': 'Illinois',
    'IN': 'Indiana',
    'KS': 'Kansas',
    'KY': 'Kentucky',
    'LA': 'Louisiana',
    'MA': 'Massachusetts',
    'MD': 'Maryland',
    'ME': 'Maine',
    'MI': 'Michigan',
    'MN': 'Minnesota',
    'MO': 'Missouri',
    'MS': 'Mississippi',
    'MT': 'Montana',
    'NC': 'North Carolina',
    'ND': 'North Dakota',
    'NE': 'Nebraska',
    'NH': 'New Hampshire',
    'NJ': 'New Jersey',
    'NM': 'New Mexico',
    'NV': 'Nevada',
    'NY': 'New York',
    'OH': 'Ohio',
    'OK': 'Oklahoma',
    'OR': 'Oregon',
    'PA': 'Pennsylvania',
    'RI': 'Rhode Island',
    'SC': 'South Carolina',
    'SD': 'South Dakota',
    'TN': 'Tennessee',
    'TX': 'Texas',
    'UT': 'Utah',
    'VA': 'Virginia',
    'VT': 'Vermont',
    'WA': 'Washington',
    'WI': 'Wisconsin',
    'WV': 'West Virginia',
    'WY': 'Wyoming'
}

##### Since most of datasets(Enrichment Data) contains only State Names instead of State&StateFIPS, which is missing in final combined COVID19 dataset.
- We have taken US States Codes & Names mapping in a list

In [166]:
USCovid19DF['StateName']=USCovid19DF['State'].map(USStateMapping)
USCovid19DF = USCovid19DF[['ReportDate', 'StateFIPS', 'State', 'StateName', 'CountyFIPS','CountyName','ConfirmCasesCount','DeathCount','Population']]
USCovid19DF['Population']=USCovid19DF['Population'].astype('Int64')
USCovid19DF['ReportDate'] = pd.to_datetime(USCovid19DF['ReportDate'], format='%Y-%m-%d')
USCovid19DF.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2397943 entries, 0 to 2397942
Data columns (total 9 columns):
 #   Column             Dtype         
---  ------             -----         
 0   ReportDate         datetime64[ns]
 1   StateFIPS          int64         
 2   State              object        
 3   StateName          object        
 4   CountyFIPS         int64         
 5   CountyName         object        
 6   ConfirmCasesCount  int64         
 7   DeathCount         int64         
 8   Population         Int64         
dtypes: Int64(1), datetime64[ns](1), int64(4), object(3)
memory usage: 185.2+ MB


##### Covid19 combined dataset improvements and make useful & easy next steps for analysis
- Added StateName column based on States Mapping list
- Rearranged columns for easy readability
- Converted Population column datatype from Float to Int
- Converted ReportDate datatype from Object to Date datatype
- Verified data and now Population&StateName added to final dataset

In [167]:
USCovid19DF.isnull().sum()

ReportDate               0
StateFIPS                0
State                    0
StateName                0
CountyFIPS               0
CountyName               0
ConfirmCasesCount        0
DeathCount               0
Population           43558
dtype: int64

##### Verified final Covid19 Dataset has any null values, observe data Population column has 43,558 blanks(due to left join) outof 2,397,943

In [168]:
USCovid19DF.to_csv('..\\..\\Data\\US_Covid_19_Dataset.csv', index=False)
USCovid19ReadDF = pd.read_csv('..\\..\\Data\\US_Covid_19_Dataset.csv')
USCovid19ReadDF.head()

ReportDate  StateFIPS State StateName  CountyFIPS             CountyName  \
0  2020-01-22          1    AL   Alabama           0  Statewide Unallocated   
1  2020-01-22          1    AL   Alabama        1001         Autauga County   
2  2020-01-22          1    AL   Alabama        1003         Baldwin County   
3  2020-01-22          1    AL   Alabama        1005         Barbour County   
4  2020-01-22          1    AL   Alabama        1007            Bibb County   

   ConfirmCasesCount  DeathCount  Population  
0                  0           0         0.0  
1                  0           0     55869.0  
2                  0           0    223234.0  
3                  0           0     24686.0  
4                  0           0     22394.0

###### Exported Python Dataframe data into csv file and stored under Data directory named: US_Covid_19_Dataset.csv
- And also read again from final dataset data into dataframe and verified all working fine for next Enrichment analysis
- We have observed that again Population datatype changed from Int to Float, which we handled in Enrichment data integration

In [169]:
USCovid19ReadDF['Population']=USCovid19ReadDF['Population'].astype('Int64')
USCovid19ReadDF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2397943 entries, 0 to 2397942
Data columns (total 9 columns):
 #   Column             Dtype 
---  ------             ----- 
 0   ReportDate         object
 1   StateFIPS          int64 
 2   State              object
 3   StateName          object
 4   CountyFIPS         int64 
 5   CountyName         object
 6   ConfirmCasesCount  int64 
 7   DeathCount         int64 
 8   Population         Int64 
dtypes: Int64(1), int64(4), object(4)
memory usage: 166.9+ MB


##### Converted Population Datatype from Float to Int, since its changed during reimport final dataset from CSV 